napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,122 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import RLock
+from typing import Dict, Iterable, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+LOGGER = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    def __init__(self):
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self.lock = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is reubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        existing_task = self._find_existing_task(layers)
+        # are the tasks unique?
+        if existing_task:
```

Comment:
Not entirely sure what you mean by unique. As the tasks are stored in a dict from `Tuple[Layer]` to `Future`, then for a particular tuple of layers we will either get one future or `None`.

```suggestion
        if existing_task := self._find_existing_task(layers):
```

Reply:
I've since discovered that Futures have a hash and are, therefore, unique. This was stemming from the inverse dict not being able to handle non-unique values. Even with a loop, we'll need them to be unique. Anyway, question resolved. I'll clean it up.